### PR TITLE
Test using dynamic container

### PR DESCRIPTION
**Pull request by jasonb5**
- Fix workflow to always build containers (utilizes cache for speedup)
- Testing always uses newest container
- Fix recursion error from `case.preview` call for E3SM

Test suite:
Test baseline:
Test namelist changes:
Test status: [bit for bit, roundoff, climate changing]

Fixes [CIME Github issue #]
User interface changes?
Update gh-pages html (Y/N)?:

### CODE REVIEW

**Comment by jasonb5** [comment]
Looks like there is still a "maximum recursion depth exceeded in comparison" error introduced by #4572 and is still specific to E3SM.

**Comment by jasonb5** [comment]
The recursion error was caused by `CIME.XML.env_batch._submit_single_job` not correctly handling the `dry_run` argument for tests.

E3SM system testing exposed the issue during it's pre-run provenance capture where it calls `case.preview_run` which calls `_submit_single_job` with the `dry_run` argument. Since the `dry_run` argument was not handled for tests it would cause `_submit_single_job` to submit the job again, causing the recursion error.

**Comment by jasonb5** [comment]
TODO
- [ ] Update docs

**Review by jgfouca — APPROVED** [review]
LGTM

File: .github/scripts/ghcr-prune.py [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+)
+
+kwargs = vars(parser.parse_args())
+
+logging.basicConfig(level=kwargs["log_level"])
+
+logger = logging.getLogger("ghcr-prune")
+
+
+class GitHubPaginate:
```
Comment (jedwards4b) [inline .github/scripts/ghcr-prune.py:37]: Can you document this a little - what is it intended to do?

File: docker/Dockerfile [hunk 2]
```diff
@@ -84,10 +77,38 @@ RUN curl -L -k -o "${PWD}/pnetcdf.tar.gz" \
             make install && \
             rm -rf "${PWD}/pnetcdf"
 
-RUN mkdir -p /root/.cime /storage/timings
+# CESM dependencies
```
Comment (jedwards4b) [inline docker/Dockerfile:80]: Is this the only place we need to update dependencies for cesm/cime?

Comment (jedwards4b) [reply docker/Dockerfile:80]: Should they be passed in as the external dependancies (eg ESMF_VERSION) are?

**Review by jedwards4b — APPROVED** [review]

File: docker/Dockerfile (hunk 2 above)
Comment (jasonb5) [reply docker/Dockerfile:80]: Yes this is the only place you need to update dependencies for cesm/cime.

I can update these to build arguments as well but I would like to **always** have the defaults specified in the Dockerfile and not the workflow.

Comment (jedwards4b) [reply docker/Dockerfile:80]: If I understand you correctly then you should not make them arguments.

**Review by jedwards4b — APPROVED** [review]
Thanks!

File: .github/scripts/ghcr-prune.py (hunk 1 above)
Comment (jasonb5) [reply .github/scripts/ghcr-prune.py:47]: Adds description to tool and docstring for class. See the ghcr-prune workflow for usage.

**Pull request merged**